File: src/ios/ScreenGuard.m
#import <Cordova/CDV.h>
#import <UIKit/UIKit.h>

@interface ScreenGuard : CDVPlugin
@end

@implementation ScreenGuard

- (void)pluginInitialize {
    [self startObservers];
}

- (void)startObservers {
    // Screenshot detection
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onScreenshot)
                                                 name:UIApplicationUserDidTakeScreenshotNotification
                                               object:nil];

    // Screen recording detection
    [[UIScreen mainScreen] addObserver:self
                            forKeyPath:@"captured"
                               options:NSKeyValueObservingOptionNew
                               context:nil];
}

- (void)onScreenshot {
    [self sendEvent:@"screenshot"];
}

- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary<NSKeyValueChangeKey,id> *)change
                       context:(void *)context {

    BOOL isCaptured = [UIScreen mainScreen].isCaptured;

    if (isCaptured) {
        [self sendEvent:@"recording_started"];
    } else {
        [self sendEvent:@"recording_stopped"];
    }
}

- (void)sendEvent:(NSString *)eventName {
    NSString *js = [NSString stringWithFormat:@"window.SCREEN_GUARD_CB && window.SCREEN_GUARD_CB('%@');", eventName];
    [self.commandDelegate evalJs:js];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    @try {
        [[UIScreen mainScreen] removeObserver:self forKeyPath:@"captured"];
    } @catch (NSException *exception) {
        // ignore
    }
}

@end

File: src/ios/ScreenGuard.swift
import Foundation
import UIKit

@objc(ScreenGuard) class ScreenGuard: CDVPlugin {

    var callbackId: String?

    @objc(start:)
    func start(command: CDVInvokedUrlCommand) {
        self.callbackId = command.callbackId

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(screenshotTaken),
            name: UIApplication.userDidTakeScreenshotNotification,
            object: nil
        )

        let screen = UIScreen.main
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(screenCapturedChanged),
            name: UIScreen.capturedDidChangeNotification,
            object: nil
        )

        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
        self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
    }

    @objc func screenshotTaken() {
        sendEvent("screenshot")
    }

    @objc func screenCapturedChanged() {
        let captured = UIScreen.main.isCaptured
        sendEvent(captured ? "recording_started" : "recording_stopped")
    }

    func sendEvent(_ event: String) {
        guard let callbackId = self.callbackId else { return }
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: event)
        result?.setKeepCallbackAs(true)
        self.commandDelegate!.send(result, callbackId: callbackId)
    }

    @objc(stop:)
    func stop(command: CDVInvokedUrlCommand) {
        NotificationCenter.default.removeObserver(self)
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
        self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
    }
}
